src/common-util/src/main/java/ee/ria/xroad/common/util/SchemaValidator.java

```diff
@@ -24,53 +24,141 @@
 
 import ee.ria.xroad.common.CodedException;
 
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.xml.sax.InputSource;
 import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+import org.xml.sax.helpers.DefaultHandler;
 
 import javax.xml.XMLConstants;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
 import javax.xml.transform.Source;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.sax.SAXSource;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.transform.stream.StreamSource;
 import javax.xml.validation.Schema;
 import javax.xml.validation.SchemaFactory;
-import javax.xml.validation.Validator;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
 import java.net.URL;
 
 /**
  * Base class for schema-based validators.
  */
 @Slf4j
-public class SchemaValidator {
+public abstract class SchemaValidator {
 
     protected static Schema createSchema(String fileName) {
-        SchemaFactory factory =
-                SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+        SchemaFactory factory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+
         try {
             URL schemaLocation = ResourceUtils.getClasspathResource(fileName);
+
             return factory.newSchema(schemaLocation);
         } catch (SAXException e) {
-            log.error("Creating schema from file '{}' failed",
-                    fileName, e);
-            throw new RuntimeException(
-                    "Unable to create schema validator", e);
+            log.error("Creating schema from file '{}' failed", fileName, e);
+
+            throw new RuntimeException("Unable to create schema validator", e);
         }
     }
 
-    protected static void validate(Schema schema, Source source,
-            String errorCode) throws Exception {
+    protected static void validate(Schema schema, Source source, String errorCode) throws Exception {
         if (schema == null) {
             throw new IllegalStateException("Schema is not initialized");
         }
 
         try {
-            Validator validator = schema.newValidator();
-            validator.validate(source);
+            validateWithSaxParser(source, schema);
```

Comment:
We are currently using the library 'xerces:xercesImpl:2.11.0.SP5' and validation with its DOM parser didn't work. When we investigated library's source code we noticed that the code related with secure processing is missing there!?

Reply:
The Validator implementation is actually Xerces 2.11.0 which does not support the mentioned properties. And Xerces is required because the Java default implementation lacks some features we currently need (https://github.com/ria-ee/X-Road/pull/17#issuecomment-287692430).

It would be possible to select the Java default implementation of the SchemaFactory using SchemaFactory.newInstance(schemaLanguage, factoryClassName, classLoader), but that would require hard-coding the internal factory class name. 

I am also somewhat concerned about the DOMSource conversion in the SchemaValidator#toInputSource since it transforms a DOMSource via a byte stream (causing XML serialization and deserialization), and that happens (at least) in signature verification which is done for every message. Have you tested possible performance implications?

Reply:
We didn't noticed performance implications, but now Java default implementation is used.